diff -r 7df563e4b7fd imap/imap.c
--- a/imap/imap.c	Tue Sep 04 17:22:59 2007 -0700
+++ b/imap/imap.c	Wed Sep 05 00:40:46 2007 -0700
@@ -1866,7 +1866,7 @@ imap_complete_hosts (char *dest, size_t 
   int matchlen;
   
   matchlen = mutt_strlen (dest);
-  for (mailbox = Incoming; mailbox && mailbox->next; mailbox = mailbox->next)
+  for (mailbox = Incoming; mailbox; mailbox = mailbox->next)
   {
     if (!mutt_strncmp (dest, mailbox->path, matchlen))
     {
